More obvious card filtering UI on term table pages

zarino opened the pull request:
Fixes #6273 by introducing a new dropdown menu that you click to switch the visible "section" of each person-card (eg: bio, contacts, social).

Fixes #5408 by attaching the new filter bar to the top of the window when scrolling.

There's space to add a "Group:" dropdown (ie: a reworking of #5349) but I've left that to be done in a different PR.

Screenshots don't really do the scrolling awesomeness justice, but we can try:

### Desktop width, no filters active

![screen shot 2016-08-10 at 14 44 09](https://cloud.githubusercontent.com/assets/739624/17556007/a95bd1ba-5f09-11e6-82b5-30af684e0316.png)

### Desktop width, filtered by search term

![screen shot 2016-08-10 at 14 50 44](https://cloud.githubusercontent.com/assets/739624/17556051/deeb8b7c-5f09-11e6-8e0b-6ed2d18bc425.png)

### Narrow device, filtered by search term, and also scrolled down to show how the filters attach to the top of the screen

(Sorry, arrows missing on the dropdown because Chrome device imitation is rubbish – they're there on the real devices)

![screen shot 2016-08-10 at 14 46 06](https://cloud.githubusercontent.com/assets/739624/17556070/f29596c2-5f09-11e6-9b60-778b93549d1e.png)

All tested and working in all the major browsers. Degrades nicely in IE8.

zarino commented on views/sass/_components.scss:
```diff
@@ -180,6 +180,21 @@ p.lead {
 }
 
+.form-control,
+input[type].form-control {
```
I'm having to overload selectors with this `input[type]` stuff because the `input[type="…"]` selectors in `_typography.scss` are just way too strong.

Bootstrap gets round this by not applying any styles to inputs by default, instead forcing you to add a `.form-control` class if you want your inputs styled. We could do the same. But I wonder whether there’s a happy medium?

Any thoughts @wrightmartin @ondenman @octopusinvitro?

zarino commented on views/term_table.erb:
```diff
@@ -107,130 +99,146 @@
                 </div>
-            </div>
 
-            <div class="grid-list" data-active-section="bio">
```
Sorry it appears I've changed lots of lines in this file – in actuality, all I did was indent it all 4 spaces further, so I could wrap it in the `<div class="js-fixed-parent">`.

zarino commented:
Would one (or more!) of @octopusinvitro @ondenman @chrismytton like to review this?

tmtmtmtm commented:
@zarino I'm assuming the big "What's in this dataset" section shouldn't still be saying "Click one of the categories to get a taste of the information we have so far"?

zarino commented:
> @zarino I'm assuming the big "What's in this dataset" section shouldn't still be saying "Click one of the categories to get a taste of the information we have so far"?

Good catch!

tmtmtmtm commented:
@zarino: generally looks great, yay! This is a *lot* better than what we currently have. I'm finding it a little bit jumpy at times though. e.g. if I'm on https://everypolitician-viewe-pr-14341.herokuapp.com/estonia/riigikogu/term-table/12.html, and scrolled to the Annely Akkermann row:

![screen shot 2016-08-10 at 16 45 58](https://cloud.githubusercontent.com/assets/57483/17560527/53e4e5a8-5f1a-11e6-9af3-9a770b2a54f2.png)

when I switch to the 'Social links' or 'Contact details' view , that row mostly shifts upwards out of sight again:

![screen shot 2016-08-10 at 16 46 24](https://cloud.githubusercontent.com/assets/57483/17560485/27c3e582-5f1a-11e6-814c-36a5dc244c53.png)

Identifiers then brings them down again:

![screen shot 2016-08-10 at 16 47 53](https://cloud.githubusercontent.com/assets/57483/17560511/443eccf4-5f1a-11e6-9d29-3e6e885e6e81.png)

zarino commented:
@tmtmtmtm: Yeah, when you switch to showing different bits of the cards, there's no guarantee the cards will be the same height (chances are, they definitely won't) so they'll move around, which can accumulate to quite a large jump when you're scrolled a long way down the list.

@tmtmtmtm: What did you expect to happen when you changed the "show details" option? Perhaps that the cards currently at the top of your view stayed at the top of your view?

tmtmtmtm commented:
@zarino Yeah — I would have expected the top of my screen to stay consistent, so if I start at 

![screen shot 2016-08-10 at 17 05 44](https://cloud.githubusercontent.com/assets/57483/17561236/071bc61c-5f1d-11e6-8cf8-0f8e2f5b94e3.png)

and switch to Social, I'd be at:

![screen shot 2016-08-10 at 17 05 55](https://cloud.githubusercontent.com/assets/57483/17561246/13cfff9a-5f1d-11e6-808c-d41055eb2a43.png)

If that's in "Eeek! Way too difficult" territory, I'd also be happy enough with it snapping to the top of those cards, like:

![screen shot 2016-08-10 at 17 06 04](https://cloud.githubusercontent.com/assets/57483/17561262/2c0939aa-5f1d-11e6-98e5-51239e232e90.png)

But the current behaviour pushes us down to the next set, which I find quite odd:

![screen shot 2016-08-10 at 17 06 30](https://cloud.githubusercontent.com/assets/57483/17561273/30ac63d8-5f1d-11e6-9567-dbf0850be1a7.png)

It's not a show-stopper, but it was quite jarring.

zarino commented:
That should be possible. I'll take a look tomorrow.

zarino commented:
@tmtmtmtm – It now attempts to maintain the visual scroll position after the "show details" select box is changed. Works in all the browsers I’ve tested it in, apart from:

* IE8 – doesn't support `position: fixed` so the "show details" select can't follow you down the page, meaning nothing to worry about here!
* IE9 – some weird flexbox CSS bug is causing the person-cards not to wrap onto multiple lines, meaning you can't scroll down the page at all – I'll try to fix that later.
* Safari on Mac – there's a few milliseconds between the CSS styles being updated and the javascript resetting the scroll position. It's noticeable that something has flashed on the screen. But I'm guessing this is better than losing your place in the list?
* Safari on iOS – Oh, fun and games!! We’re falling foul of [this "feature" in Safari iOS](http://stackoverflow.com/questions/29001977 ), which pulls you up to the top of the page when you focus the `position: fixed` select box. Another thing to fix tomorrow.

zarino commented:
I've fixed the iOS bug, so this is ready for review. @chrismytton could you review the JavaScript? And @wrightmartin the Sass, maybe?

chrismytton commented on public/javascript/ep-card-filter.js:
```diff
@@ -194,9 +199,45 @@ $(document).ready(function(){
 
-    $('[data-section-toggle]').on('click', function(){
-      var section = $(this).attr('data-section-toggle');
+    $('.js-show-facet').on('change', function(){
```
This line can be simplified by using `.change(handler)`, which is a [shortcut for `.on('change', handler)`](https://api.jquery.com/change/).

zarino replied:
Is it not preferable to maintain consistency with all the other event bindings elsewhere, which all use `.on()`?

chrismytton replied:
I can only see two others in this file, but I guess that does make it the majority 😉  

I agree, consistency is best for this pull request, we can change to the shorthand methods in a future PR 👍

zarino replied:
> we can change to the shorthand methods in a future PR

Interesting. My vote would always be for using `.on()` everywhere – although I'd be interested to hear the argument for shorthand methods, aside from being five characters shorter.

chrismytton commented on public/javascript/main.js:
```diff
@@ -272,13 +268,166 @@
+      }
+    }
+
+    updateChildPositionAbsolute = function updateChildPositionAbsolute($parent, $child, $spacer){
```
I think you're missing a `var` at the start of this line.

chrismytton commented:
@zarino Couple of comments on the js, but generally looking fine to me! 👍  🎉

zarino commented:
@chrismytton @tmtmtmtm: I’ve rebased this on top of the latest master. Ran the tests locally in my vm, and they all passed, and it all looks and works fine when I poke at it in the browser. So I think we’re ready to merge this in, if you’re happy with it?

tmtmtmtm commented:
The travis failures here look like they're from https://github.com/everypolitician/everypolitician-data/pull/18231#issuecomment-252073921

I've fixed that in https://github.com/everypolitician/everypolitician-data/pull/18255, and will try restarting travis here when that gets merged.

zarino commented:
> Is this fairly easy to fix, or is it likely to be quite involved? I think my expected behaviour would be that a search would always bring the first matches for that search to the top

Shouldn't be hard to fix. I'll take a look now.

> making the search box much more prominent like this also makes #10276 even more important now

Good spot! I will fix that at the same time.

zarino commented:
@tmtmtmtm – Those two issues have been fixed now. Modified keys (eg: Ctrl-A) now work as you’d expect in the search box (fixing #10276) and searching takes you back to the top of the card list, so you don't get stranded at the bottom.

tmtmtmtm commented:
Woot! This is a huge improvement. Thanks @zarino and @octopusinvitro
